stingray/deadtime/filters.py

```diff
@@ -0,0 +1,161 @@
+import warnings
+import copy
+
+import numpy as np
+import numpy.random as ra
+from astropy import log
+from astropy.logger import AstropyUserWarning
+
+from ..utils import njit
+from ..events import EventList
+
+
+def _paralyzable_dead_time(event_list, dead_time):
+    mask = np.ones(len(event_list), dtype=bool)
+    dead_time_end = event_list + dead_time
+    bad = dead_time_end[:-1] > event_list[1:]
+    # Easy: paralyzable case. Here, events coming during dead time produce
+    # more dead time. So...
+    mask[1:][bad] = False
+
+    return event_list[mask], mask
+
+
+@njit()
+def _nonpar_core(event_list, dead_time_end, mask):
+    for i in range(1, len(event_list)):
+        if (event_list[i] < dead_time_end[i - 1]):
+            dead_time_end[i] = dead_time_end[i - 1]
+            mask[i] = False
+    return mask
+
+
+def _non_paralyzable_dead_time(event_list, dead_time):
+    event_list_dbl = (event_list - event_list[0]).astype(np.double)
+    dead_time_end = event_list_dbl + np.double(dead_time)
+    mask = np.ones(event_list_dbl.size, dtype=bool)
+    mask = _nonpar_core(event_list_dbl, dead_time_end, mask)
+    return event_list[mask], mask
+
+
+class DeadtimeFilterOutput(object):
+    uf_events = None
+    is_event = None
+    deadtime = None
+    mask = None
+    bkg = None
+
+
+def filter_for_deadtime(event_list, deadtime, bkg_ev_list=None,
+                        dt_sigma=None, paralyzable=False,
+                        additional_data=None, return_all=False):
+    """Filter an event list for a given dead time.
+
+    Parameters
+    ----------
+    ev_list : array-like or class:`stingray.events.EventList`
+        The event list
+    deadtime: float
+        The (mean, if not constant) value of deadtime
+
+    Other Parameters
+    ----------------
+    bkg_ev_list : array-like
+        A background event list that affects dead time
+    dt_sigma : float
```

Comment:
I think it's worth being specific here that this will result in dead time being draws from a normal distribution with `mean = deadtime` and `variance = dt_sigma`

Reply:
Done